subprojects/file-watching/src/main/java/org/gradle/internal/watch/registry/FileWatcherUpdater.java

```diff
@@ -25,11 +25,14 @@
 
 public interface FileWatcherUpdater extends SnapshotHierarchy.SnapshotDiffListener {
     /**
-     * Changes the must watch directories, e.g. when the same daemon is used on a different project.
+     * Changes the project root directories, e.g. when the same daemon is used on a different project.
+     *
+     * The project root directories are used by hierarchical watchers to minimize the number of watched roots
+     * by watching the project roots instead of watching directories inside.
      *
      * @throws WatchingNotSupportedException when the native watchers can't be updated.
      */
-    void updateMustWatchDirectories(Collection<File> updatedWatchDirectories);
+    void updateProjectRootDirectories(Collection<File> updatedProjectRootDirectories);
```

Comment:
Why does this take a collection instead of a single `File`?

Reply:
We started out with a collection for some reason. I can move to a single file for now to make the code simpler and we can support a collection if we need at some point. WDYT?

Reply:
One thing a collection support is not having any project root directory. Though we currently don't have a use-case for that.

Reply:
Right. And maybe even if we had such a thing, we could keep track of that by using `null` instead of a collection. In any case, I think a single value should be good. We could call it `currentProjectRoot` or `mostRecentlySeenProjectRoot` or something.

Reply:
I think `projectRootDirectory` is still good as a name.